### PR TITLE
Add link to analytics options in submenu and make all settings sections linkable

### LINKED ISSUE #5190: Restore Analytics admin menu item to link to jump to expanded Analytics drawer on settings screen
## Feature description

In #5155 for #5013 we moved the Analytics screen to a drawer under the Advanced section on the Settings screen. As part of this, the admin menu item for Analytics was removed. However, users are accustomed to seeing this item. It would be helpful for users to retain this link but have it deep link to the settings screen like via `/wp-admin/admin.php?page=amp-options#analytics-options-drawer`

For that matter, each of the drawers should be deep-linkable. If the ID for the drawer is the location target/hash, then it should be scrolled into view, expanded, and focused when the component first mounts. This will be helpful to link users to Plugin Suppression, for example. This should also be the case for non-drawers as well. For example, there should be a way to link to the mobile redirection toggle so that it will scroll into view and be focused.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation brief

* 

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

### PR DESCRIPTION
**Pull request by johnwatkins0**
## Summary

Fixes #5190  

In #5155 we removed the Analytics menu page and placed the analytics option on the settings screen. This restores that menu item with a link to the Analytics section of the menu page.

Additional changes:
- Passed the URL hash into the root settings screen component so that it can be used to scroll to sections after the UI loads
- Moved the drawers on the settings screen up to the root settings screen component so that all the handling of IDs/anchors can be in one file
- Created an `AnalyticsOptionsSubmenu` service class, which is a stripped-down version of the class that was removed in #5155. It inserts a link into the AMP admin menu in the second position. 
- Created a `get_menu_slug` method in the `OptionsMenu` class for consumption in `AnalyticsOptionsSubmenu`. 

## Checklist

- [x] My pull request is addressing an open issue (please create one otherwise).
- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

File: assets/src/settings-page/index.js [hunk 1]
```diff
@@ -94,8 +95,11 @@ ErrorNotice.propTypes = {
 
 /**
  * Settings page application root.
+ *
+ * @param {Object} props Component props.
+ * @param {string} props.section The initially focused section.
```
Comment (westonruter) [inline assets/src/settings-page/index.js]: Maybe `focusedSection` would be more clear?

File: assets/src/settings-page/index.js [hunk 2]
```diff
@@ -120,6 +124,25 @@ function Root() {
...
+		}
+
+		if ( ! section ) {
+			return;
+		}
+
+		const focusedSection = document.getElementById( section );
+
+		if ( focusedSection ) {
+			focusedSection.scrollIntoView();
```
Comment (westonruter) [inline assets/src/settings-page/index.js]: Would smooth scrolling be desirable? 

```suggestion
			focusedSection.scrollIntoView( { behavior: 'smooth' } );
```

Comment (westonruter) [reply assets/src/settings-page/index.js]: Alternatively, for accessibility, I think what is needed is to actually `focus()` on the first focusable element. 

So this should be looking for the first `button` or `input` that is inside the `focusedSection` and then `focus()` on it. This will automatically scroll _and_ add the necessary keyboard focus.

File: src/Admin/AnalyticsOptionsSubmenu.php [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+		global $submenu;
+
+		$old_menu = $submenu[ $this->parent_menu_slug ];
+
+		// The link should be the second menu item.
+		$new_menu = array_merge(
+			[ $old_menu[0] ],
+			[
+				[
+					__( 'AMP Analytics Options', 'amp' ),
```
Comment (westonruter) [inline src/Admin/AnalyticsOptionsSubmenu.php]: This needs to be shortened to just `Analytics`:

image

File: src/Admin/AnalyticsOptionsSubmenu.php [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+					'manage_options',
+					menu_page_url( $this->parent_menu_slug, false ) . '#analytics-options',
+				],
+			],
+			array_slice( $old_menu, 1 )
+		);
+
+		$submenu[ $this->parent_menu_slug ] = array_values( $new_menu ); // phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited
```
Comment (westonruter) [inline src/Admin/AnalyticsOptionsSubmenu.php]: Should not `$submenu` be modified with `add_submenu_page()` rather than doing this and the above?

File: assets/src/settings-page/index.js [hunk 5]
```diff
@@ -161,7 +222,7 @@ domReady( () => {
 		render( (
 			<ErrorContextProvider>
 				<Providers>
-					<Root optionsRestPath={ OPTIONS_REST_PATH } />
+					<Root section={ global.location.hash.replace( /^#/, '' ) } />
```
Comment (westonruter) [inline assets/src/settings-page/index.js]: I think this can be improved… namely, if I am already on the AMP Settings screen and click the admin menu item, then I am jumped down the screen without the section expanding. Therefore, it would be preferable if there was also a [`hashchange`](https://developer.mozilla.org/en-US/docs/Web/API/Window/hashchange_event) event that would re-render this tree whenever it fires. This will ensure that the section always reflects what is currently in the `location.hash`.

**Review by westonruter — CHANGES_REQUESTED** [review]

File: src/Admin/AnalyticsOptionsSubmenu.php (hunk 4 above)
Comment (johnwatkins0) [reply src/Admin/AnalyticsOptionsSubmenu.php]: Ah, you're right. Updated in 0299a03c0. Didn't realize it would be so simple to link to the parent menu page with a hash appended.

File: src/Admin/AnalyticsOptionsSubmenu.php (hunk 3 above)
Comment (johnwatkins0) [reply src/Admin/AnalyticsOptionsSubmenu.php]: Done in b28aa90a4

File: assets/src/settings-page/index.js (hunk 1 above)
Comment (johnwatkins0) [reply assets/src/settings-page/index.js]: Done in e41c971d9

File: assets/src/settings-page/index.js (hunk 2 above)
Comment (johnwatkins0) [reply assets/src/settings-page/index.js]: Done in `e41c971d9` with `select` elements also included

File: assets/src/settings-page/index.js [hunk 6]
```diff
@@ -94,8 +95,13 @@ ErrorNotice.propTypes = {
 
 /**
  * Settings page application root.
+ *
+ * @param {Object} props Component props.
+ * @param {string} props.focusedSection The initially focused focusedSection.
```
Comment (westonruter) [inline assets/src/settings-page/index.js]:
```suggestion
```

File: assets/src/settings-page/index.js [hunk 7]
```diff
@@ -153,6 +221,9 @@ function Root() {
 		</>
 	);
 }
+Root.propTypes = {
+	focusedSection: PropTypes.string,
+};
```
Comment (westonruter) [inline assets/src/settings-page/index.js]:
```suggestion
```

File: assets/src/settings-page/index.js [hunk 8]
```diff
@@ -94,8 +95,13 @@ ErrorNotice.propTypes = {
 
 /**
  * Settings page application root.
+ *
+ * @param {Object} props Component props.
+ * @param {string} props.focusedSection The initially focused focusedSection.
  */
 function Root() {
+	const [ focusedSection, setFocusedSection ] = useState( global.location.hash.replace( /^#/, '' ) );
```
Comment (westonruter) [inline assets/src/settings-page/index.js:123]: `setFocusedSection` isn't yet used. It will be set in a `hashchange` event handler?

Comment (johnwatkins0) [reply assets/src/settings-page/index.js:123]: Yup, in 754aadd8572de9744af0f99953b077c3bded9bf9

File: assets/src/settings-page/index.js (hunk 5 above)
Comment (johnwatkins0) [reply assets/src/settings-page/index.js]: I did this in 754aadd8572de9744af0f99953b077c3bded9bf9 but it was surprisingly complicated to get it to work in all the scenarios (initial load of the page, changing the hash in the URL location bar, clicking the Analytics link in the menu), and I'm not sure how I feel about how much code I had to add. It's working, though.

File: assets/src/settings-page/index.js [hunk 9]
```diff
@@ -120,6 +144,41 @@ function Root() {
...
+	/**
+	 * Resets the focused element state when the hash changes on the page.
+	 */
+	useEffect( () => {
+		const hashChangeCallback = ( event = null ) => {
+			if ( event ) {
+				event.preventDefault();
+			}
+
+			// Ensure this runs after state upates.
```
Comment (pierlon) [inline assets/src/settings-page/index.js]:
```suggestion
			// Ensure this runs after state updates.
```

File: assets/src/settings-page/index.js [hunk 10]
```diff
@@ -132,13 +191,48 @@ function Root() {
 				saveOptions();
 			} }>
 				<TemplateModes />
-				<h2>
+				<h2 id="advanced-settings">
 					{ __( 'Advanced Settings', 'amp' ) }
 				</h2>
 				<MobileRedirection />
```
Comment (pierlon) [inline assets/src/settings-page/index.js]: Why not put the ID for the section here instead of inside of the component?

File: assets/src/settings-page/index.js [hunk 11]
```diff
@@ -132,13 +191,48 @@ function Root() {
...
+					heading={ (
+						<h3>
+							{ __( 'Supported Templates', 'amp' ) }
+						</h3>
+					) }
+					hiddenTitle={ __( 'Supported templates', 'amp' ) }
+					id="supported-templates"
+					initialOpen={ 'supported-templates' === focusedSection }
+				>
+					<SupportedTemplates />
```
Comment (pierlon) [inline assets/src/settings-page/index.js:209]: Same question as above.

File: tests/e2e/utils/visit-admin-page-with-hash.js [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+ */
+import { join } from 'path';
+
+/**
+ * WordPress dependencies
+ */
+import { getPageError } from '@wordpress/e2e-test-utils';
+
+function createURLWithHash( WPPath, query = '', hash = '' ) {
+	const url = new URL( 'http://localhost:8890' );
```
Comment (pierlon) [inline tests/e2e/utils/visit-admin-page-with-hash.js]: This URL should be available via the environment variable [`WP_BASE_URL`](https://github.com/ampproject/amp-wp/blob/6f92f315964cb5b6629edff2893968bf4ae27b9b/package.json#L131), so no need to hard code it.

File: tests/e2e/utils/visit-admin-page-with-hash.js [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+/**
+ * Visits admin page with hash option; if user is not logged in then it logging in it first, then visits admin page.
+ *
+ * @param {string} adminPath String to be serialized as pathname.
+ * @param {string} query String to be serialized as query portion of URL.
+ * @param {string} hash URL hash.
+ */
+export async function visitAdminPageWithHash( adminPath, query, hash = null ) {
```
Comment (pierlon) [inline tests/e2e/utils/visit-admin-page-with-hash.js]: If defaulting `hash` to `null` is still wanted I'd recommend updating the docblock to include `null`, otherwise I'd recommend defaulting to an empty string:

```suggestion
export async function visitAdminPageWithHash( adminPath, query, hash = '' ) {
```

File: tests/php/src/Admin/AnalyticsOptionsSubmenuTest.php [hunk 14]
```diff
@@ -0,0 +1,100 @@
+<?php
+/**
+ * Tests for AnalyticsOptionsSubmenu.
+ *
+ * @package AmpProject\AmpWP\Tests
+ */
+
+namespace AmpProject\AmpWP\Tests\Admin;
+
+use AMP_Options_Manager;
```
Comment (pierlon) [inline tests/php/src/Admin/AnalyticsOptionsSubmenuTest.php]: Not being used.

```suggestion
```

File: tests/php/src/Admin/AnalyticsOptionsSubmenuTest.php [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+use AmpProject\AmpWP\Admin\GoogleFonts;
+use AmpProject\AmpWP\Admin\OptionsMenu;
+use AmpProject\AmpWP\Admin\ReaderThemes;
+use AmpProject\AmpWP\Admin\RESTPreloader;
+use WP_UnitTestCase;
+
+/**
+ * Tests for AnalyticsOptionsSubmenu.
+ *
+ * @group options-menu
```
Comment (pierlon) [inline tests/php/src/Admin/AnalyticsOptionsSubmenuTest.php:21]: I've recently noticed that the Travis job that runs the PHP tests with code coverage has been emitting a [lot of warnings](https://travis-ci.org/github/ampproject/amp-wp/jobs/716788170#L847), most of which are due to incorrect uses of the `@covers` and `@uses` annotations. I've managed to fix them in https://github.com/ampproject/amp-wp/pull/5201/commits/cfa46c486453e39ec330371fe60944dccdb57f17 but I'd like to prevent them from occurring in the future, so let's start using them the way they're meant to be :smiley:.

/cc @westonruter.

```suggestion
 * @group options-menu
 * @coversDefaultClass \AmpProject\AmpWP\Admin\AnalyticsOptionsSubmenu
```

File: tests/php/src/Admin/AnalyticsOptionsSubmenuTest.php [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+			new ReaderThemes(),
+			new RESTPreloader()
+		);
+		$this->instance              = new AnalyticsOptionsSubmenu( $this->options_menu_instance );
+	}
+
+	/**
+	 * Test register.
+	 *
+	 * @see AnalyticsOptionsSubmenu::register()
```
Comment (pierlon) [inline tests/php/src/Admin/AnalyticsOptionsSubmenuTest.php]:
```suggestion
	 * @covers :register()
```

File: tests/php/src/Admin/AnalyticsOptionsSubmenuTest.php [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+	 */
+	public function test_register() {
+		$this->instance->register();
+		$this->assertEquals( 99, has_action( 'admin_menu', [ $this->instance, 'add_submenu_link' ] ) );
+	}
+
+	/**
+	 * Test add_submenu_link.
+	 *
+	 * @covers AnalyticsOptionsSubmenu::add_submenu_link()
```
Comment (pierlon) [inline tests/php/src/Admin/AnalyticsOptionsSubmenuTest.php]:
```suggestion
	 * @covers ::add_submenu_link()
```

**Review by pierlon — CHANGES_REQUESTED** [review]

File: assets/src/settings-page/index.js (hunk 10 above)
Comment (johnwatkins0) [reply assets/src/settings-page/index.js]: Yeah, I like having all the IDs at this level. Updated in 5e7b5de0734c92a5ab0f163656ee6abf21af96e6

File: assets/src/settings-page/index.js (hunk 11 above)
Comment (johnwatkins0) [reply assets/src/settings-page/index.js:209]: In this case the ID is set just above on the `AMPDrawer` component.

File: tests/e2e/utils/visit-admin-page-with-hash.js (hunk 13 above)
Comment (johnwatkins0) [reply tests/e2e/utils/visit-admin-page-with-hash.js]: Updated in 92ec1f3e6

File: tests/e2e/utils/visit-admin-page-with-hash.js (hunk 12 above)
Comment (johnwatkins0) [reply tests/e2e/utils/visit-admin-page-with-hash.js]: Updated in 92ec1f3e6

File: tests/e2e/specs/admin/reader-theme-carousel.js [hunk 18]
```diff
@@ -6,6 +6,7 @@ import { visitAdminPage } from '@wordpress/e2e-test-utils';
 describe( 'AMP settings screen reader themes carousel', () => {
 	beforeEach( async () => {
 		await visitAdminPage( 'admin.php', 'page=amp-options' );
+		await page.addStyleTag( { content: 'html {scroll-behavior: auto !important;}' } );
```
Comment (johnwatkins0) [inline tests/e2e/specs/admin/reader-theme-carousel.js:9]: Figuring this out was rough. Puppeteer kept failing with "Node is either not visible or not an HTMLElement", and I found this buried as the last answer in one of the lower-ranking StackOverflow results. Essentially, the `scroll-behavior: smooth` CSS rule I added to the setting screen messes up Puppeteer.

**Comment by johnwatkins0** [comment]
Re-requesting review after addressing feedback and getting tests to pass.

**Comment by jwold** [comment]
I like the change. It took me a moment to catch on that the analytics menu was a jump down, but the rest of the items were separate pages. Apart from that I think it's a good example of still making Analytics findable, but keeping it inside the main settings page. 

Screen Capture on 2020-08-13 at 13-11-07

File: src/Admin/Polyfills.php [hunk 19]
```diff
@@ -128,6 +128,10 @@ public function register_shimmed_scripts( $wp_scripts ) {
 	 * @param WP_Styles $wp_styles The WP_Styles instance for the current page.
 	 */
 	public function register_shimmed_styles( $wp_styles ) {
+		if ( version_compare( get_bloginfo( 'version' ), '5.4', '<' ) ) {
```
Comment (johnwatkins0) [inline src/Admin/Polyfills.php:131]: Unrelated change: Noticed some undesirable CSS effects when working in 5.2 locally, so this will force the polyfilled wp-components styles. Note this only runs on AMP screens.

**Comment by pierlon** [comment]
@googlebot I consent.

**Comment by pierlon** [comment]
Some tests were modifying the global `$submenu` without restoring it's original value, so I've fixed that in https://github.com/ampproject/amp-wp/pull/5216/commits/70e517b6bf3f5d82c95c0c745d8820d9de3eaed3. I've also updated `OptionsMenuTest::test_add_menu_items` to perform the correct assertions.

**Comment by pierlon** [comment]
On a larger viewport the scroll to the Analytics section doesn't seem to have much have an effect:

**Comment by johnwatkins0** [comment]
> On a larger viewport the scroll to the Analytics section doesn't seem to have much have an effect:


That's true, but it's just because the analytics section is last on the page so it can't scroll any further than that. I suppose the workarounds would be to move that section up or add a bit of extra padding to the bottom of the page.

**Comment by pierlon** [comment]
Could we focus the drawer when selected by hash? Similar to how GitHub does it, where if I [link to a comment](https://github.com/ampproject/amp-wp/pull/5216#issuecomment-673900209) it adds a border around it:

image

**Comment by westonruter** [comment]
> Could we focus the drawer when selected by hash? Similar to how GitHub does it, where if I [link to a comment](https://github.com/ampproject/amp-wp/pull/5216#issuecomment-673900209) it adds a border around it:

I'd say that would be an optional enhancement. GitHub styles this via:

```css
:target .timeline-comment {
    border-color: #2188ff;
    box-shadow: 0 0 0 0.2em #c8e1ff;
}
```

However, the drawers depend on `box-shadow` to achieve their effect. If we do any such highlighting, I'd say perhaps we go with an initial faint yellow that fades out (e.g. via a CSS animation).

But yeah, I'd say this is not really necessary. After all, the first focusable element will have the outline. I suppose the the alternative would be remove the `:not(.components-panel__body-toggle)` so that the drawer button gets the border which would have a similar effect to what GitHub is doing:

image

```diff
--- a/assets/src/settings-page/index.js
+++ b/assets/src/settings-page/index.js
@@ -105,7 +105,7 @@ function scrollFocusedSectionIntoView( focusedSectionId ) {
 		return;
 	}
 
-	const firstInput = focusedSectionElement.querySelector( 'input, select, textarea, button:not(.components-panel__body-toggle)' );
+	const firstInput = focusedSectionElement.querySelector( 'input, select, textarea, button' );
 
 	if ( firstInput ) {
 		firstInput.focus();
```

**Comment by westonruter** [comment]
I noticed that when going to `/wp-admin/admin.php?page=amp-options#reader-themes-drawer` it doesn't cause the Reader theme drawer to open initially.

**Comment by johnwatkins0** [comment]
a856c71c5 adds scrolling to the reader themes drawer if it's on the page (note I changed that HTML ID from `reader-themes-drawer` to `reader-themes`). I also removed that `:not(.components-panel__body-toggle)` in the selector, so the toggle button will always be focused for these drawers.

Another thing, I noticed that `element.scrollIntoView({behavior: smooth})` seemed to work much more reliably than `focusElement.focus({behavior: smooth})` so I updated it to use `scrollIntoView` on the main element and then focus the focus element without smooth scroll.

**Comment by westonruter** [comment]
> `focusElement.focus({behavior: smooth})`

Oh, I didn't even the `focus()` method as supporting `{behavior:smooth}`: https://developer.mozilla.org/en-US/docs/Web/API/HTMLOrForeignElement/focus

So, good to switch to `scrollIntoView` 👍

File: assets/src/settings-page/style.css [hunk 20]
```diff
@@ -361,13 +364,13 @@ li.error-kept {
 	border-bottom-width: 1px;
 }
 
-#reader-themes-drawer {
+#reader-themes {
```
Comment (westonruter) [inline assets/src/settings-page/style.css:367]: Thanks for making this change. I was going to suggest it but I didn't want to nit pick… too much. 😊

**Comment by johnwatkins0** [comment]
> Oh, I didn't even the `focus()` method as supporting `{behavior:smooth}`:

Huh, you're right. I could've sworn I learned this was a thing recently, but I'm no longer able to find it 🤷‍♂️  No wonder it didn't work.

**Comment by westonruter** [comment]
For some reason deeplinking to `#reader-themes` is not causing scrolling to occur. If I try navigating in a new tab to `#analytics` I get scrolled down, but not if I go to `#reader-themes`. If, however, I put the scroll logic in a 100ms timeout then the Reader Themes will scroll into view. It's strange because the Reader Themes section does exist in the DOM and it is being rendered. Not a blocker since this section is near the top anyway and it is still getting expanded, but it is still a mystery.

File: tests/e2e/specs/admin/reader-theme-carousel.js (hunk 18 above)
Comment (westonruter) [reply tests/e2e/specs/admin/reader-theme-carousel.js:9]: Good find.

**Review by westonruter — APPROVED** [review]

**Review by pierlon — APPROVED** [review]
